ament_cmake_export_libraries fails when CMAKE_INSTALL_LIBDIR is set

It looks like we don't have an issue template on this repo. I used Gemini to find this and generate an MRE, but I've reviewed it and think it's valid.

## Recommended actions:

I think we should fix it in Lyrical, and get rid of the old-style CMake variable macros going forwards.

* Make ament_export_libraries respect CMAKE_INSTALL_LIBDIR, and backport that to ROS Lyrical
* Deprecate and remove ament_export_libraries in ROS Rolling, and remove it after ROS Makoa

## Context

* [ROS Lyrical build of behaviortree_cpp](https://build.ros2.org/view/Lbin_uR64/job/Lbin_uR64__behaviortree_cpp__ubuntu_resolute_amd64__binary/10/consoleFull)
* [ROS Lyrical build of auto_apms_behavior_tree_core](https://build.ros2.org/view/Lbin_uR64/job/Lbin_uR64__auto_apms_behavior_tree_core__ubuntu_resolute_amd64__binary/21/console)

behaviortree_cpp uses `ament_export_libraries`

* https://github.com/BehaviorTree/BehaviorTree.CPP/blob/4630e066f842f4f7509d3362b43b651106876cd6/cmake/ament_build.cmake#L27

The binary build for `behaviortree_cpp` sets `CMAKE_INSTALL_LIBDIR` to `lib/x86_64-linux-gnu`

```
18:05:30 	cd .obj-x86_64-linux-gnu && DEB_PYTHON_INSTALL_LAYOUT=deb PKG_CONFIG=/usr/bin/pkg-config cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DFETCHCONTENT_FULLY_DISCONNECTED=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DBUILD_TESTING:BOOL=OFF -DCMAKE_INSTALL_PREFIX=/opt/ros/lyrical -DAMENT_PREFIX_PATH=/opt/ros/lyrical -DCMAKE_PREFIX_PATH=/opt/ros/lyrical -DBUILD_TESTING=OFF ..
```

`auto_apms_behavior_tree_core` fails to find_package `behaviortree_cpp`:

```
23:50:45 -- Found behaviortree_cpp: 4.9.1 (/opt/ros/lyrical/share/behaviortree_cpp/cmake)
23:50:45 CMake Error at /opt/ros/lyrical/share/behaviortree_cpp/cmake/ament_cmake_export_libraries-extras.cmake:48 (message):
23:50:45   Package 'behaviortree_cpp' exports the library 'behaviortree_cpp' which
23:50:45   couldn't be found
23:50:45 Call Stack (most recent call first):
23:50:45   /opt/ros/lyrical/share/behaviortree_cpp/cmake/behaviortree_cppConfig.cmake:41 (include)
23:50:45   CMakeLists.txt:15 (find_package)
```

# Begin AI bug report:

## Description
When a package built with a custom `CMAKE_INSTALL_LIBDIR` (such as `lib/x86_64-linux-gnu`) exports a library using `ament_export_libraries()`, downstream packages fail to configure via `find_package()`.

The generated file `ament_cmake_export_libraries-extras.cmake` hardcodes the search directory to `lib` using an absolute/relative path with `NO_DEFAULT_PATH`. This prevents CMake from locating libraries installed in arch-specific or customized directories.

---

## Steps to Reproduce

1. Create a workspace containing `behaviortree_cpp` (which exports libraries using `ament_export_libraries`) and a minimal downstream package `btcpp_mre`.

2. Use the minimal package configuration below.

<details>
<summary>CMakeLists.txt</summary>

```cmake
cmake_minimum_required(VERSION 3.8)
project(btcpp_mre)

find_package(ament_cmake REQUIRED)
find_package(behaviortree_cpp REQUIRED)

ament_package()
```
</details>

<details>
<summary>package.xml</summary>

```xml
<?xml version="1.0"?>
<package format="3">
  <name>btcpp_mre</name>
  <version>0.0.1</version>
  <description>Minimal Reproducible Example for ament_cmake_export_libraries issue</description>
  <maintainer email="test@test.com">test</maintainer>
  <license>MIT</license>

  <buildtool_depend>ament_cmake</buildtool_depend>
  <depend>behaviortree_cpp</depend>

  <export>
    <build_type>ament_cmake</build_type>
  </export>
</package>
```
</details>

3. Build the workspace specifying a customized `CMAKE_INSTALL_LIBDIR` (default on many multi-arch Debian/Ubuntu systems):
   ```bash
   colcon build --packages-select behaviortree_cpp btcpp_mre --cmake-args -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu
   ```

### Output / Error Trace
```
Starting >>> behaviortree_cpp
Finished <<< behaviortree_cpp [42.9s]
Starting >>> btcpp_mre

CMake Error at /workspaces/ros2/install/behaviortree_cpp/share/behaviortree_cpp/cmake/ament_cmake_export_libraries-extras.cmake:48 (message):
  Package 'behaviortree_cpp' exports the library 'behaviortree_cpp' which
  couldn't be found
```

---

## Root Cause
In `ament_cmake_export_libraries-extras.cmake.in`, the path is hardcoded:
```cmake
      find_library(
        _lib NAMES "${_library}"
        PATHS "${@PROJECT_NAME@_DIR}/../../../lib"
        NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH
      )
```
This fails to find libraries located under `${@PROJECT_NAME@_DIR}/../../../lib/x86_64-linux-gnu`.

---

## Proposed Solution
Respect the configured `CMAKE_INSTALL_LIBDIR` dynamically in the template generator of `ament_cmake_export_libraries`:

```diff
diff --git a/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in b/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
index bd47853..1974733 100644
--- a/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
+++ b/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
@@ -37,9 +37,13 @@ if(NOT _exported_libraries STREQUAL "")
     if(NOT IS_ABSOLUTE "${_library}")
       # search for library target relative to this CMake file
       set(_lib "NOTFOUND")
+      set(_install_libdir "@CMAKE_INSTALL_LIBDIR@")
+      if(_install_libdir STREQUAL "")
+        set(_install_libdir "lib")
+      endif()
       find_library(
         _lib NAMES "${_library}"
-        PATHS "${@PROJECT_NAME@_DIR}/../../../lib"
+        PATHS "${@PROJECT_NAME@_DIR}/../../../${_install_libdir}"
         NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH
       )
```

--- a/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
+++ b/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
@@ -37,9 +37,13 @@ if(NOT _exported_libraries STREQUAL "")
     if(NOT IS_ABSOLUTE "${_library}")
       # search for library target relative to this CMake file
       set(_lib "NOTFOUND")
+      set(_install_libdir "@CMAKE_INSTALL_LIBDIR@")
+      if(_install_libdir STREQUAL "")
+        set(_install_libdir "lib")
+      endif()
       find_library(
         _lib NAMES "${_library}"
-        PATHS "${@PROJECT_NAME@_DIR}/../../../lib"
+        PATHS "${@PROJECT_NAME@_DIR}/../../../${_install_libdir}"
         NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH
       )
 

--- a/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
+++ b/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
@@ -39,7 +39,11 @@ if(NOT _exported_libraries STREQUAL "")
       set(_lib "NOTFOUND")
       find_library(
         _lib NAMES "${_library}"
-        PATHS "${@PROJECT_NAME@_DIR}/../../../@CMAKE_INSTALL_LIBDIR@"
+        PATHS
+        # This is for packages that install libraries to destinations using GNUInstallDirs variables.
+        "${@PROJECT_NAME@_DIR}/../../../@CMAKE_INSTALL_LIBDIR@"
+        # Historically ROS packages have installed libraries to this hardcoded lib directory.
+        "${@PROJECT_NAME@_DIR}/../../../lib"
         NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH
       )
 

--- a/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
+++ b/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries-extras.cmake.in
@@ -37,13 +37,9 @@ if(NOT _exported_libraries STREQUAL "")
     if(NOT IS_ABSOLUTE "${_library}")
       # search for library target relative to this CMake file
       set(_lib "NOTFOUND")
-      set(_install_libdir "@CMAKE_INSTALL_LIBDIR@")
-      if(_install_libdir STREQUAL "")
-        set(_install_libdir "lib")
-      endif()
       find_library(
         _lib NAMES "${_library}"
-        PATHS "${@PROJECT_NAME@_DIR}/../../../${_install_libdir}"
+        PATHS "${@PROJECT_NAME@_DIR}/../../../@CMAKE_INSTALL_LIBDIR@"
         NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH
       )
 

--- a/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries_package_hook.cmake
+++ b/ament_cmake_export_libraries/cmake/ament_cmake_export_libraries_package_hook.cmake
@@ -18,6 +18,7 @@
 # as well as linker flags
 
 # generate and register extra file for libraries
+include(GNUInstallDirs)
 set(_generated_extra_file
   "${CMAKE_CURRENT_BINARY_DIR}/ament_cmake_export_libraries/ament_cmake_export_libraries-extras.cmake")
 configure_file(

